sciris/sc_utils.py

```diff
@@ -1553,8 +1554,61 @@ def flattendict(inputdict=None, basekey=None, subkeys=None, complist=None, keyli
             complist.append(comp)
     return complist, keylist
 
+def suggest(user_input, valid_inputs, threshold=4, die=False):
+    """
+    Return suggested item
+
+    Returns item with lowest Levenshtein distance, where case substitution and stripping
+    whitespace are not included in the distance. If there are ties, then the additional operations
+    will be included
+
+    :param user_input: String with user's input
+    :param valid_inputs: List/collection of valid strings
+    :param threshold: Maximum number of edits required for an option to be suggested
+    :param die: If True, an informative error will be raised (to avoid having to implement this in the calling code)
+    :return: Suggested string. Returns None if no suggestions with edit distance less than threshold were found. This helps to make
+             suggestions more relevant.
 
+    Example usage:
+
+    >>> suggest('foo',['Foo','Bar'])
+    'Foo'
+    >>> suggest('foo',['FOO','Foo'])
+    'Foo'
+    >>> suggest('foo',['Foo ','boo'])
+    'Foo '
 
+    """
+
+    if isstring(valid_inputs):
+        valid_inputs = [valid_inputs]
+    else:
+        valid_inputs = list(valid_inputs)
+
+    distance = np.zeros(len(valid_inputs))
+    cs_distance = np.zeros(len(valid_inputs))
+    # We will switch inputs to lowercase because we want to consider case substitution a 'free' operation
+    # Similarly, stripping whitespace is a free operation. This ensures that something like
+    # 'foo ' will match 'Foo' ahead of 'boo '
+    for i, s in enumerate(valid_inputs):
+        distance[i] = Levenshtein.distance(user_input, s.strip().lower())
+        cs_distance[i] = Levenshtein.distance(user_input, s.strip())
+
+    if sum(distance==min(distance)) > 1:
+        # If there is a tie for the minimum distance, use the case sensitive comparison
+        distance = cs_distance
+
+    suggestion = valid_inputs[np.argmin(distance)]
+
+    if min(distance) > threshold:
+        if die:
+            raise Exception('"%s" not found' % user_input)
+        else:
+            return None
+    elif die:
+        raise Exception('"%s" not found - did you mean "%s"?' % (user_input, suggestion))
+    else:
+        return suggestion
```

Comment:
Would also suggest an optional kwarg `fulloutput=False` (or a better name!), which if `True` returns an odict of `{valid_input[i]:distance[i]}` pairs